File: packages/sdk-python/proofsdk/pack_readiness.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any


def select_pack_readiness(pack: Mapping[str, Any]) -> dict[str, Any] | None:
    if pack.get("pack_completeness_profile") or pack.get("pack_completeness_status"):
        return {
            "source": "pack_scoped",
            "profile": pack.get("pack_completeness_profile"),
            "status": pack.get("pack_completeness_status"),
            "pass_count": pack.get("pack_completeness_pass_count"),
            "warn_count": pack.get("pack_completeness_warn_count"),
            "fail_count": pack.get("pack_completeness_fail_count"),
        }

    if pack.get("completeness_profile") or pack.get("completeness_status"):
        return {
            "source": "bundle_aggregate",
            "profile": pack.get("completeness_profile"),
            "status": pack.get("completeness_status"),
            "pass_count": pack.get("completeness_pass_count"),
            "warn_count": pack.get("completeness_warn_count"),
            "fail_count": pack.get("completeness_fail_count"),
        }

    return None

File: sdks/typescript/src/utils/pack_readiness.ts
import type {
  PackManifest,
  PackReadinessSummary,
  PackSummaryResponse,
} from "../types.js";

type PackLike = PackSummaryResponse | PackManifest;

function asOptionalStatus(
  value: unknown,
): PackReadinessSummary["status"] | undefined {
  return value === "pass" || value === "warn" || value === "fail"
    ? value
    : undefined;
}

function asOptionalNumber(value: unknown): number | undefined {
  return typeof value === "number" ? value : undefined;
}

export function selectPackReadiness(
  pack: PackLike,
): PackReadinessSummary | null {
  if (pack.pack_completeness_profile || pack.pack_completeness_status) {
    return {
      source: "pack_scoped",
      profile: pack.pack_completeness_profile,
      status: pack.pack_completeness_status,
      passCount: pack.pack_completeness_pass_count,
      warnCount: pack.pack_completeness_warn_count,
      failCount: pack.pack_completeness_fail_count,
    };
  }

  if (pack.completeness_profile || pack.completeness_status) {
    const legacyRecord = pack as Record<string, unknown>;
    return {
      source: "bundle_aggregate",
      profile: pack.completeness_profile,
      status: asOptionalStatus(legacyRecord.completeness_status),
      passCount: asOptionalNumber(legacyRecord.completeness_pass_count),
      warnCount: asOptionalNumber(legacyRecord.completeness_warn_count),
      failCount: asOptionalNumber(legacyRecord.completeness_fail_count),
    };
  }

  return null;
}
